### DISCUSSION TITLE
FLAC Support?

### DISCUSSION THREAD

**Discussion by RollingStar** [OpenRCT2/OpenRCT2#15641] [Feature requests] [answered]
https://github.com/RollingStar/RCT-Music-Patch

FLAC has better support by modern applications. I didn't do any work with css1 because it looks like others are getting to the bottom of sound routines on their own. Waiting for knowledge on separating out those sound effects saves me some time. I expect something like QuickBMS or a custom, small program can separate out the sound effects.

**Comment by IntelOrca** [comment]
We will need an open source FLAC encoder (decoder at least) for OpenRCT2 itself too, so that it can read the music files that are compressed. Preferably a single C file like lodepng (but might have to use libFLAC).

I was thinking of adding a command line tool to OpenRCT2 itself for copying original data files to OpenRCT2's own install directory where compression would then take place, e.g.

`openrct2.exe install E:\`

**Comment by zsilencer** [comment]
Is this only for saving disk space?  I could see using encoders for custom music but the program requires installation of Rollercoaster Tycoon 2 to run.  I'm not sure users would want to modify the installed files just to save ~200MB, but if an installer is created that copies those files to a new OpenRCT2 directory then it might be a good idea.

Before this can be considered, there is a bit of decompiling to be done so that the audio can be safely put into our own functions, such as the vehicle sounds routines etc.

**Comment by IntelOrca** [comment]
@zsilencer We'll probably have a NSIS installer for Windows which allow the user to specify the location of their original RCT2 files which are then copied. I was also thinking of built in functionality to do this too with OpenRCT2 via command line, particular for linux users who like that kind of thing. Both could compress the music files during the copy.

BTW, nice work on the audio decompiling. Won't be long now until we can change over to software mixed SDL2.

**Comment by RollingStar** [comment]
> I was thinking of adding a command line tool to OpenRCT2 itself

The Windows registry stores the install location, even for GOG installs and probably for Steam installs.

`reg query "HKLM\SOFTWARE\Wow6432Node\Infogrames\RollerCoaster Tycoon 2 Setup" /v Path`

`reg query "HKLM\SOFTWARE\Infogrames\RollerCoaster Tycoon 2 Setup" /v Path`

> Is this only for saving disk space? ... I'm not sure users would want to modify the installed files just to save ~200MB, but if an installer is created that copies those files to a new OpenRCT2 directory then it might be a good idea.

Exactly. OpenRCT2 is moving away from sharing files with RCT2. I hope a smaller file size means a faster loading time. Even on an SSD, the loading time is nearly identical between OpenRCT2 and RCT2. (I don't think the startup process is decompiled yet.) I expect some [smart optimizations](https://support.steampowered.com/kb_article.php?ref=7388-QPFN-2491) with how many [read calls](https://github.com/IntelOrca/OpenRCT2/wiki/Performance) are made to the operating system could reduce load times. I also have a hunch that smaller files have got to be easier to load somewhere, at some point in the program.

The other benefit is better metadata. GOG already provides a soundtrack (that's worse than mine), so there's clearly interest. Jonny Heykens never made something called "Heykens' Serenade". That's what other people called it after the fact. He wrote it as "Serenade, op. 21". And by contacting the sound artist for RCT2 I confirmed that my track listing was correct - so we can move beyond "Rock Style" and onto "Airtime Rock".

Anyone listening to one of the non-original tracks can now know who wrote them, and are just a Google search away from other tracks by the same person. Examples include Charles-Marie Widor's Toccata (Organ Style) and Johann Strauss II's Tales from the Vienna Woods (Fairground Organ Style 5).

I also put them into what a track order would probably look like - Title Theme at the start, then basically into the order they're listed in the credits. I classify the 1 hour, 40 minutes of music as disc 1 in the metadata, and start a sound effects disc with the crowd noise and circus fanfare. That's where the rest of the effects will go when they're separated out. Anyone dumping the RCT2 WAVEs into a folder so they can listen to them won't have the track order or the track titles. I scrobble to [Last FM](http://www.last.fm/), so correct titles are important.

**Comment by karst** [comment, marked as answer]
Introduced by #17217
